pkg/pipeline/stream/destination.go

```diff
@@ -51,23 +50,30 @@ func (n *DestinationNode) Run(ctx context.Context) (err error) {
 	if err != nil {
 		return cerrors.Errorf("could not open destination connector: %w", err)
 	}
-	defer func() {
-		// wait for acker node to receive all outstanding acks, time out after
-		// 1 minute or right away if the context is already canceled.
-		waitCtx, cancel := context.WithTimeout(ctx, time.Minute)
-		defer cancel()
-		n.AckerNode.Wait(waitCtx)
 
-		// teardown will kill the plugin process
-		tdErr := n.Destination.Teardown(connectorCtx)
-		if tdErr != nil {
+	// lastPosition stores the position of the last successfully processed record
+	var lastPosition record.Position
+	// openMsgTracker tracks open messages until they are acked or nacked
+	var openMsgTracker OpenMessagesTracker
+	defer func() {
+		stopErr := n.Destination.Stop(connectorCtx, lastPosition)
```

Comment:
You got everything right, except the conclusion 😄 First thing to note is that `Run` receives a context that will be canceled if any node in the pipeline stops running because of an error. Second thing - when we call `n.Destination.Open` we pass it a context, if that context gets canceled the stream will get closed immediately. Also any calls to the connector need to happen with an open context (e.g. `Stop` and `Teardown`), otherwise it's considered that the call was canceled (I expect gRPC wouldn't even deliver a call with a closed context).

So essentially we create a new context `connectorCtx` to be sure it will stay open until the node actually stops running, so we can gracefully stop the connector.

Reply:
Thanks for the clarification! 

> First thing to note is that `Run` receives a context that will be canceled if any node in the pipeline stops running because of an error.

I believe it would be nice to have this in a comment in the code, it's useful.

Reply:
https://github.com/ConduitIO/conduit/blob/1793c51112c16c1fa46b25bfadeee4aecff1e4a1/pkg/pipeline/stream/node.go#L30-L60

Reply:
What I believe would be useful in a comment (be it the one above or wherever it makes sense), is the note that the context will be cancelled if any pipeline node stops running. But it's a nitpick level.: )

Reply:
I added a clarification 👍